### PR TITLE
Enums

### LINKED ISSUE #18: derive Bounded, Enum and BoundedEnum for Variant
I think we can derive Bounded, Enum and BoundedEnum for Variant, an it would be pretty useful, as for normal sum types we already have `Generic.Rep.{Bounded,Enum}`.

### PR DESCRIPTION
**Pull request by cryogenian**
Fixes #18

### CODE REVIEW

**Comment by cryogenian** [comment]
@natefaubion Does it look good?

**Comment by natefaubion** [comment]
I will take a look at this ASAP. Thanks for the ping.

File: src/Data/Variant.purs [hunk 1]
```diff
@@ -217,6 +223,211 @@ instance eqVariant ∷ (R.RowToList r rl, VariantTags rl, VariantEqs rl) ⇒ Eq
...
+  ⇒ Bounded (Variant r)
+  where
+    top =
+      let
+        prx = RLProxy ∷ RLProxy rl
+        tags = variantTags prx
+        dicts = variantBounded prx
+        zipped = L.zip tags dicts
+        hd = fromJust "top" $ L.head $ L.sortBy (\a b → compare (T.fst b) (T.fst a)) zipped
+      in unsafeCoerce { type: T.fst hd, value: (T.snd hd).top }
```
Comment (natefaubion) [inline src/Data/Variant.purs]: I don't think there's a reason to `zip` and `sortBy` here. `RowToList` is already sorted, so this should just be a matter of using `head` and `last`. Can we have an annotated `VariantRep a -> Variant r` coercion?

File: src/Data/Variant.purs [hunk 2]
```diff
@@ -217,6 +223,211 @@ instance eqVariant ∷ (R.RowToList r rl, VariantTags rl, VariantEqs rl) ⇒ Eq
...
+
+    bottom =
+      let
+        prx = RLProxy ∷ RLProxy rl
+        tags = variantTags prx
+        dicts = variantBounded prx
+        zipped = L.zip tags dicts
+        hd = fromJust "bottom" $ L.head $ L.sortBy (\a b → compare (T.fst a) (T.fst b)) zipped
+      in unsafeCoerce { type: T.fst hd, value: (T.snd hd).bottom }
```
Comment (natefaubion) [inline src/Data/Variant.purs]: Same as `top`.

File: src/Data/Variant.purs (hunk 1 above)
Comment (natefaubion) [reply src/Data/Variant.purs]: I know it's more code, but I'd like to avoid making extra traversals/allocations over the lists in general. Since this is typeclass dispatch it needs to be as efficient as possible, even if that means writing special purpose loops over lists.

File: src/Data/Variant.purs [hunk 3]
```diff
@@ -217,6 +223,211 @@ instance eqVariant ∷ (R.RowToList r rl, VariantTags rl, VariantEqs rl) ⇒ Eq
...
+
+        VariantRep arep = unsafeCoerce a
+
+        tagIx ∷ Int
+        tagIx = fromJust "enum" $ L.findIndex (\(T.Tuple s _) → arep.type == s) zipped
+
+        pack = fromJust "enum" $ L.index zipped tagIx
+
+        dict ∷ BoundedEnumDict VariantCase
+        dict = T.snd $ T.snd pack
```
Comment (natefaubion) [inline src/Data/Variant.purs]: I have the same concern about this implementation. This zips the lists and performs 3 linear lookups. This could be done with a single traversal of the list.

File: src/Data/Variant.purs [hunk 4]
```diff
@@ -217,6 +223,211 @@ instance eqVariant ∷ (R.RowToList r rl, VariantTags rl, VariantEqs rl) ⇒ Eq
...
+
+        totalTags = L.length tags
```
Comment (natefaubion) [inline src/Data/Variant.purs]: Same as `pred`, except I think this performs an additional traversal for length.

File: src/Data/Variant.purs [hunk 5]
```diff
@@ -217,6 +223,211 @@ instance eqVariant ∷ (R.RowToList r rl, VariantTags rl, VariantEqs rl) ⇒ Eq
...
+
+      go ∷ Int → L.List (T.Tuple String (BoundedEnumDict VariantCase)) → Int
+      go acc = case _ of
+        L.Cons (T.Tuple k dict) tl
+          | k == arep.type →
+              acc + dict.fromEnum arep.value
+          | otherwise →
+              let c = dict.cardinality
+              in go (acc + c) tl
+        _ → unsafeCrashWith "Data.Variant: impossible boundedEnum"
```
Comment (natefaubion) [inline src/Data/Variant.purs]: Similar issues. Unnecessary sorting, we don't need the additional zip, `tagIx` and `pack` are unused.

File: src/Data/Variant/Internal.purs [hunk 6]
```diff
@@ -109,14 +114,30 @@ lookup
...
+  M.Just a → a
+  _ → unsafeCrashWith $ "Data.Variant: impossible `" <> name <> "`"
+
+lookup'
+  ∷ ∀ a
+  . String
+  → L.List String
+  → L.List a
+  → M.Maybe a
+lookup' tag = go
```
Comment (natefaubion) [inline src/Data/Variant/Internal.purs]: I don't think these changes are necessary given other feedback. I'd like to avoid the `Just` box since we just immediately remove it.

File: src/Data/Variant.purs [hunk 7]
```diff
@@ -217,6 +223,211 @@ instance eqVariant ∷ (R.RowToList r rl, VariantTags rl, VariantEqs rl) ⇒ Eq
...
+          | dict.cardinality > ix  → case dict.toEnum ix of
+              M.Just a → M.Just $ unsafeCoerce { type: k, value: a }
+              M.Nothing → M.Nothing -- this is impossible in fact
+          | otherwise → go (ix - dict.cardinality) tl
+        _ → M.Nothing
+
+
+
+
+
```
Comment (natefaubion) [inline src/Data/Variant.purs]: Lots of unnecessary whitespace. Single empty lines are used everywhere else.

**Comment by cryogenian** [comment]
@natefaubion I think I've addressed you comments. Could you take a look again, please?

**Comment by natefaubion** [comment]
Thanks! I've cleaned this up a bit, and moved the lookups to `Internal`.
